tree-walk.go

```diff
@@ -28,51 +28,62 @@ type treeWalkResult struct {
 	end   bool
 }
 
-// listDir - lists all the entries at a given prefix, takes additional params as filter and leaf detection.
-// filter is required to filter out the listed entries usually this function is supposed to return
-// true or false.
-// isLeaf is required to differentiate between directories and objects, this is a special requirement for XL
-// backend since objects are kept as directories, the only way to know if a directory is truly an object
-// we validate if 'xl.json' exists at the leaf. isLeaf replies true/false based on the outcome of a Stat
-// operation.
-func (xl xlObjects) listDir(bucket, prefixDir string, filter func(entry string) bool, isLeaf func(string, string) bool) (entries []string, err error) {
-	for _, disk := range xl.getLoadBalancedQuorumDisks() {
-		if disk == nil {
-			continue
-		}
-		entries, err = disk.ListDir(bucket, prefixDir)
-		if err != nil {
-			// For any reason disk was deleted or goes offline, continue
-			// and list form other disks if possible.
-			if err == errDiskNotFound || err == errFaultyDisk {
+// "listDir" function of type listDirFn returned by getListDir() - explained below.
+type listDirFn func(bucket, prefixDir string, filter func(entry string) bool) (entries []string, err error)
+
+// Returns function "listDir" of the type listDirFn.
+// isLeaf - is used by listDir function to check if an entry is a leaf or non-leaf entry.
+// disks - used for doing disk.ListDir(). FS passes single disk argument, XL passes a list of disks.
+func getListDir(isLeaf func(string, string) bool, disks ...StorageAPI) listDirFn {
+	// listDir - lists all the entries at a given prefix, takes additional filter function as parameter.
```

Comment:
everywhere we are passing the disks slice as is and functions take care of skipping the disk if it is `nil` hence better to be consistent, so the onus of filtering of nil disks is never on the caller.